### ISSUE TITLE
High Rate of Request Drops in TGI Stress Test: Seeking Clarity and Debugging Guidance

### ISSUE THREAD

**Issue by martinigoyanes** [huggingface/text-generation-inference#1733]
# Setup Characteristics
- Kubernetes Cluster deployment
- 1 A100 GPU with 80GB RAM
- 12 CPU with 32 GB RAM
- TGI version: 1.4.4
- TGI config:
              MAX_INPUT_LENGTH: "8000"
              MAX_TOTAL_TOKENS: "8512"
              MAX_BATCH_PREFILL_TOKENS: "8000"
              MAX_CONCURRENT_REQUESTS: "1024"
              LOG_LEVEL: "DEBUG"

# Short Introduction
I'm currently stress-testing TGI on my available hardware setup, specifically focusing on pushing its limits with the Mistral7b as a benchmark model. However, I'm encountering a significant number of dropped requests during this process. I'm seeking clarity on why these drops are occurring since I have not seen much documentation around ``dropped`` vs ``overloaded`` errors.

# What am I doing? 
I am flooding TGI with requests using [Locust](https://locust.io) (a load testing tool). Concretely I am sending 60 ``generate`` requests every 10 seconds during 10 minutes. Each request has ``input_length=1000`` and ``max_new_tokens=7000``.

# What behavior do I expect?
I expect TGI to either run out of memory (overloading the model) because the RAM in my GPU is not enough to handle such load or in other case, TGI successfully processes the requests.

# What behavior am I seeing? 
I am seeing more than 90% of requests being dropped when looking at the monitoring graphs in Grafana dashboard.
![Screenshot 2024-04-12 at 14 54 20](https://github.com/huggingface/text-generation-inference/assets/58919465/49df40c5-55b4-4cc1-8668-d5f0cbfac59b)

# What have I done
I have enabled debug logs by setting ``LOG_LEVEL=DEBUG`` when spawning ``text-generation-launcher``.  I have explored the issue in two ways: looking at the rust code and looking at the logs. 

### 1. Review code base
While looking at the code I found where the ``dropped`` requests are handled and where the Prometheus metrics are exported. The relevant functions are in ``infer.rs::send_responses()`` and ``infer.rs::filter_send_generations()``. Located here: https://github.com/huggingface/text-generation-inference/blob/9d8f21cace66e8593bc559174de0eed3ecdab6a2/router/src/infer.rs#L602

```
/// Send one or multiple `InferStreamResponse` to Infer for all `entries`
/// and filter entries
filter_send_generations(generations: Vec<Generation>, entries: &mut IntMap<u64, Entry>)
        ...
        // Send generation responses back to the infer task
        // If the receive an error from the Flume channel, it means that the client dropped the
        // request and we need to stop generating hence why we unwrap_or(true)
        let stopped = send_responses(generation, entry).map_err(|err| {
            tracing::error!("Entry response channel error.");
            metrics::increment_counter!("tgi_request_failure", "err" => "dropped");
            err
        }).unwrap_or(true);
...
/// Send responses through the `entry` response channel
fn send_responses(
    generation: Generation,
    entry: &Entry,
) -> Result<bool, Box<SendError<Result<InferStreamResponse, InferError>>>> {
    // Return directly if the channel is disconnected
    if entry.response_tx.is_closed() {
        metrics::increment_counter!("tgi_request_failure", "err" => "dropped");
        return Ok(true);
    }
    ...
```

However, my Rust experience is very limited, and I found it very unclear to know why exactly the request is dropped. Looks like there are two things than can happen inside the Flume channel: 
1. Either, the request connection is closed as we see in ``entry.response_tx.is_closed()`` and the function terminates.
2. Or, something else further down the line errors out while generation and it has to be stopped.

From my understanding, the ``text-generation-router`` (Rust server) is not the one "failing" but rather the ``text-generation-server`` (Python inference server), and the client used by ``text-generation-router`` sees requests are dropped by the inference server. However, it is hard to see information about this since ``log-level`` is hardcoded to ``INFO`` in python inference server, as we can see here:  https://github.com/huggingface/text-generation-inference/blob/1b2670c823e5232a4e81c16314ac6d4dc8e59cff/launcher/src/main.rs#L469
```
    // Process args
    let mut shard_args = vec![
        "serve".to_string(),
        model_id,
        "--uds-path".to_string(),
        uds_path,
        "--logger-level".to_string(),
        "INFO".to_string(),
        "--json-output".to_string(),
    ];
```
It would be nice if this parameter was tweakable from an environment variable.

### 2. Review logs

After reviewing the code, I went on an adventure to figure out the ``DEBUG`` level logs from ``text-generation-router``. Based on what I had seen in the code, I was specially looking for things such as: ``"Entry response channel error."`` since this is what the tracer logs when ``dropped`` is exported to the prometheus metric ``tgi_request_failure``. Unfourtunately, I could not find a single line in the logs about this, which is making me feel very lost. All I could found is logs about connection errors. This is an extract from the logs:
```
[2m2024-04-12T09:29:33.116146Z[0m [34mDEBUG[0m [1mConnection[0m[1m{[0m[3mpeer[0m[2m=[0mClient[1m}[0m[2m:[0m [2mh2::codec::framed_write[0m[2m:[0m [2m/usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/h2-0.3.25/src/codec/framed_write.rs[0m[2m:[0m[2m213:[0m send [3mframe[0m[2m=[0mData { stream_id: StreamId(81795) }
[2m2024-04-12T09:29:33.116151Z[0m [34mDEBUG[0m [1mConnection[0m[1m{[0m[3mpeer[0m[2m=[0mClient[1m}[0m[2m:[0m [2mh2::codec::framed_write[0m[2m:[0m [2m/usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/h2-0.3.25/src/codec/framed_write.rs[0m[2m:[0m[2m213:[0m send [3mframe[0m[2m=[0mData { stream_id: StreamId(81795), flags: (0x1: END_STREAM) }
[2m2024-04-12T09:29:33.135099Z[0m [34mDEBUG[0m [2mhyper::server::server::new_svc[0m[2m:[0m [2m/usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.28/src/server/server.rs[0m[2m:[0m[2m779:[0m connection error: connection closed before message completed
[2m2024-04-12T09:29:33.138102Z[0m [34mDEBUG[0m [2mhyper::server::server::new_svc[0m[2m:[0m [2m/usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/hyper-0.14.28/src/server/server.rs[0m[2m:[0m[2m779:[0m connection error: connection closed before message completed
```

### 3. Tweak params
I have also tried changing the ``MAX_CONCURRENT_REQUESTS`` from 128 to: 256, 512 and 1024. However that seems like a parameter that affects the model being ``Overloaded`` or not, but it does not have much impact on dropped requests. Check below:
<img width="854" alt="Screenshot 2024-04-11 at 12 46 15" src="https://github.com/huggingface/text-generation-inference/assets/58919465/45de076c-776c-41ea-84ea-c6ccefe4915a">

# Main Questions
1. When and the conditions that make a request to be dropped
2. A good practice / approach to debug TGI in issues like this. 

### Extra Information
More Grafana dashboards
![Screenshot 2024-04-12 at 14 52 49](https://github.com/huggingface/text-generation-inference/assets/58919465/6604ec99-10cf-4a1f-ae83-0afacbc58c3f)

If you are interested this is the command I am executing when running locust:
```
locust -f src/locustfile.py -L DEBUG --autostart --autoquit 120 --config-users config_users.json --users 60 --spawn-rate 1 --run-time 10m --wait-time 10 FixedInputUser

# config_users.json
{
  "user_class_name": "FixedInputUser",
  "weight": 1,
  "tasks": ["generate_text"],
  "host": "https://mistral7b.clustername.com",
  "input_length": 1000,
  "fixed_output_length": 7000
}

# generate_text pseudocode

payload = {
    "inputs": "Lorem ipsum...",
    "max_new_tokens": 7000,
    "best_of": 1,
    "decoder_input_details": False,
    "details": True,
    "do_sample": True,
    "frequency_penalty": 0.1,
    "max_new_tokens": 20,
    "repetition_penalty": 1.03,
    "return_full_text": False,
    "seed": None,
    "stop": [],
    "temperature": 0.5,
    "top_k": 10,
    "top_n_tokens": 0,
    "top_p": 0.95,
    "truncate": None,
    "typical_p": 0.95,
    "watermark": False
}

client.post(url, data=json.dumps(payload), catch_response=True, verify=False) 
```

**Comment by martinigoyanes** [comment]
Hello @OlivierDehaene and thanks a lot for such a fast response!

Correct me if I am wrong, but what is happening to my experiment is that my queue is not saturated, meaning all requests are being batched and sent for inference, but the downstream inference server is dropping the connection at ``tcp`` level. Right?

So essentially, the problem is between the rust client in ``router/client/src/client.rs`` which sends the batch to the inference server in ``server/text_generation_server/server.py``.  For some reason, the TCP connection between those two is ``dropped``,  right?

Why would such a thing ever happen if both components are interacting in ``localhost``? Could this be a TCP timeout from the Rust client to the inference server because the inference is taking TOO long? 

Anyhow, how do you recommend debugging such issues? From what I have seen the logging level for the Python inference server is hard-coded to ``INFO``: 
https://github.com/huggingface/text-generation-inference/blob/1b2670c823e5232a4e81c16314ac6d4dc8e59cff/launcher/src/main.rs#L469

By the way, thank you for pointing me to k6, I will definitely check it out!

**Comment by Narsil** [comment]
>  For some reason, the TCP connection between those two is dropped, right?

No `dropped` means your *client* is dropping the requests, since the client closed the connection, TGI also closes it.

> Why would such a thing ever happen if both components are interacting in localhost? Could this be a TCP timeout from the Rust client to the inference server because the inference is taking TOO long?

No it's not a timeout from rust, it's a timeout from your python code.
Having used locust for a long time before switching to k6, I really recommend to switch to k6, locust is giving super misrepresentative results and because it's using python it's actually quite bad at saturating things.

`overloaded` is also not an error on TGI side, it's a mecanism of backpressure, making sure you cannot give the hug of death to a tgi server. It's controlled by `max_connection_requests`.
Basically instead of miserably becoming unresponsive under load, TGI's backpressure starts answering `overloaded` (which is super cheap) to avoid starting answering a user on a request that's going to take minutes to process (looking like unresponsive).

You can deactivate it if yo want no backpressure by making MAX_CONNECTION_REQUESTS super high. Not really recommended imho, but you can :)

**Comment by martinigoyanes** [comment]
Thanks a lot, now I understand what and why is happening! 

Keep up the great work you are doing with TGI! @Narsil @OlivierDehaene
